rclcpp/doc/multi_threaded_executor_redesign.rst

```diff
@@ -0,0 +1,62 @@
+======================================
+:code:`MultiThreadedExecutor` Redesign
+======================================
+
+:code:`MultiThreadedExecutor` may attempt to execute one executable object multiple times. 
+This occurs because an executable object can be identified as ready in multiple threads before it has been executed, and each of these threads will then try to execute it.
+To fix this, a general solution should do the following:
+
+* Work for all the types of executable objects: timers, subscriptions, services, clients, and waitables
+* Avoid spurious wake ups (which can be costly)
+
+Summary of discussion so far
+----------------------------
+
+Broadly speaking, two approaches were discussed:
+
+1. **Handle no data in the executor**: If more than one thread is trying to execute the same object and that object is consumed after it is executed once, check for the executable before trying to execute it:
+
+    .. code-block:: c++
+
+        void execute() {
+          if(buffer->has_data()) {
+            auto data = buffer->consume_data();
+            run_any_callback(data);
+          } else {
+            // do nothing
+          }
+        }
+
+   :Advantages: 
+     * simplicity
+     * extends interface in a way that generalizes well for all other types of executable objects (i.e., timers)
+   :Disadvantages: 
+     * spurious wake up calls
+     * needs to be documented well so every class that executes :code:`AnyExecutable` objects is thread-safe and doesn't hide a race condition
+
+
+2. **Take the data**: When an executable is ready, take it and store it so that it can be executed when execute is called:
+
+    .. code-block:: c++
+
+        void take_data(std::shared_ptr<void> & data) {
+          data = buffer->consume_data();
+        }
+
+        void execute(std::shared_ptr<void> & data) {
+          run_any_callback(data_);
+        }
+
+   :Advantages:
+     * no spurious wake ups
```

Comment:
As commented above, this **may** avoid spurious wake ups.
But for example, it doesn't fix the problem for timers (though, it does fix the problem in all waitables were "taking data" has a meaning).

I still like the idea of somehow "taking the data" early.
I don't think that "taking data" is conceptually correct, more on that in a different comment.

For the entities where "taking data" doesn't' have sense, the problem can be fixed by making them mutually exclusive with themselves (one thread cannot wait or execute the entity if another thread is executing it).

Reply:
> For the entities where "taking data" doesn't' have sense, the problem can be fixed by making them mutually exclusive with themselves (one thread cannot wait or execute the entity if another thread is executing it).

Should this be another solution listed? Would this work for all entity types? If so, could you briefly describe its implementation?

Reply:
> Should this be another solution listed?

Distinguishing between entities that can be executed concurrently (e.g. you can take data from a subscription from different threads) from the ones that cannot sounds like right to me.

> Would this work for all entity types?

In entities like `timer`, waiting on it to be ready on one thread while scheduling it for execution in another thread is a problem.
For other entities (e.g. subscriptions), that would only worsen performance.

> If so, could you briefly describe its implementation?

We could have a flag or something that differentiates between "concurrent" and "non-concurrent" entities.

We could create a [mutually exclusive callback group](https://github.com/ros2/rclcpp/blob/0e0a6a495c452bdf3526a0b7bf457e4186bb344d/rclcpp/include/rclcpp/callback_group.hpp#L45) for each timer instead of putting them in the [default callback group](https://github.com/ros2/rclcpp/blob/0e0a6a495c452bdf3526a0b7bf457e4186bb344d/rclcpp/include/rclcpp/node_interfaces/node_base.hpp#L92).
Currently when an entity on a ME callback group is being executed, all the entities in the group (including the one scheduled for execution) cannot be waited on in another thread neither be executed.

Reply:
>We could create a mutually exclusive callback group for each timer instead of putting them in the default callback group.

How would that help? It breaks the existing expectation of users which is that callbacks are not called concurrently unless they put them in different callback groups (the same as ROS 1).

Reply:
> How would that help? It breaks the existing expectation of users which is that callbacks are not called concurrently unless they put them in different callback groups (the same as ROS 1).

AFAIU:
- Callbacks from different callback groups **can** be executed concurrently.
- Callbacks from the same Reentrant Callback Group (RCG) **can** be executed concurrently.
- Callbacks from the same Mutually Exclusive Callback Group (MECG) **cannot** be executed concurrently.

For entities that aren't reentrant (e.g. timers) we could do the following:
- By default each timer is added to a new MECG that only contains that timer, and that MECG will not be used for other entities later.
- The user can pass a custom MECG. Passing a RCG is not allowed.

I think that doesn't break user assumptions. If it does break them, I'm probably miss-understanding how callback groups work.

This might not be the ideal way of implementing non-reentrant entities, maybe having an ad-hoc flag is a better idea.

Reply:
Why are timers not reentrant? You should be able to have a timer in a reentrant callback group and have timer callbacks being called concurrently, if that's what you desire.

If you put each timer in its own MECG, then it can be run concurrently with subscription callbacks (assuming the user did not specify a custom CG for either), which is different from ROS 1, where the callbacks are mutually exclusive by default (even between subscriptions and timers).

Putting ROS 1 aside, if you right now create a timer and a subscription, and then spin in a multi-threaded executor they are mutually exclusive, your proposed change would change that behavior, breaking their expectation.

Either way, I don't see how this fixes the problem, but that's likely due to me loosing context on the conversation...

Reply:
> Why are timers not reentrant? You should be able to have a timer in a reentrant callback group and have timer callbacks being called concurrently, if that's what you desire.

We have a race condition when trying to wait on timers in a thread while another thread has scheduled its execution, and sometimes the timer is executed twice in a row ([here a failure in CI](https://ci.ros2.org/view/nightly/job/nightly_osx_release/1712/testReport/projectroot/test/test_multi_threaded_executor/), but it's almost impossible to reproduce the failure, it's extremely flaky).
There's a patch in the executor trying to avoid the race condition, but I think the patch suffers from a race condition too as explained [here](https://github.com/ros2/rclcpp/pull/1241#issuecomment-670530455).
The approach proposed in https://github.com/ros2/rclcpp/pull/1241 doesn't solve the problem and I proposed a custom workaround in https://github.com/ros2/rclcpp/pull/1275.

I think that having the concept of entities that cannot be executed reentrantly could also solve the problem, and I think that's particular suitable for timers.

> If you put each timer in its own MECG, then it can be run concurrently with subscription callbacks (assuming the user did not specify a custom CG for either), which is different from ROS 1, where the callbacks are mutually exclusive by default (even between subscriptions and timers).

Sorry, I thought that the default callback group was reentrant and not ME ([code here](https://github.com/ros2/rclcpp/blob/01d6f52e3281f6df2c191690527f9208006a5c50/rclcpp/src/rclcpp/node_interfaces/node_base.cpp#L150)). The change I proposed wouldn't be acceptable.